synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,243 @@ def enumerate_leaves(node, depth):
+    """
+
+    def drop_from_cache(self) -> None:
+        """Request the entry be dropped from the cache.
+
+        Note: This should call `.remove_from_list()` on all `_ListNodes`
+        referencing it.
+        """
+        ...
+
+
+P = TypeVar("P", bound=_CacheEntry)
+
+
+class _ListNode(Generic[P]):
+    """A node in a doubly linked list, with an (optional) weak reference to a
+    cache entry.
+
+    The weak reference should only be `None` for the root node.
+    """
+
+    # We don't use attrs here as in py3.6 you can't have `attr.s(slots=True)`
+    # and inherit from `Generic` for some reason
+    __slots__ = [
+        "cache_entry",
+        "prev_node",
+        "next_node",
+    ]
+
+    def __init__(self, cache_entry: Optional[P] = None) -> None:
+        self.cache_entry = cache_entry
+        self.prev_node: Optional[_ListNode[P]] = self
+        self.next_node: Optional[_ListNode[P]] = self
+
+    @staticmethod
+    def insert_after(
+        cache_entry: P, root: "_ListNode[P]", clock: Clock
+    ) -> "_ListNode[P]":
+        """Create a new list node that is placed after the given root node."""
+        node = _ListNode(cache_entry)
+        node.move_after(root, clock)
+        return node
+
+    def remove_from_list(self):
+        """Remove this node from the list."""
+        if self.prev_node is None or self.next_node is None:
+            # We've already been removed from the list.
+            return
+
+        prev_node = self.prev_node
+        next_node = self.next_node
+
+        prev_node.next_node = next_node
+        next_node.prev_node = prev_node
+
+        # We set these to None so that we don't get circular references,
+        # allowing us to be dropped without having to go via the GC.
+        self.next_node = None
+        self.prev_node = None
+
+        self.cache_entry = None
+
+    def move_after(self, root: "_ListNode", clock: Clock):
+        """Move this node from its current location in the list to after the
+        given node.
+        """
+        # We assert that both this node and the root node is still "alive".
+        assert self.prev_node
+        assert self.next_node
+        assert root.prev_node
+        assert root.next_node
+
+        # Remove self from the list
+        prev_node = self.prev_node
+        next_node = self.next_node
+
+        prev_node.next_node = next_node
+        next_node.prev_node = prev_node
+
+        # Insert self back into the list, after root
+        prev_node = root
+        next_node = root.next_node
+
+        self.prev_node = prev_node
+        self.next_node = next_node
+
+        prev_node.next_node = self
+        next_node.prev_node = self
+
+    def get_cache_entry(self) -> Optional[P]:
+        """Get the cache entry, returns None if this is the root node (i.e.
+        cache_entry is None) or if the entry has been dropped.
+        """
+        return self.cache_entry
+
+
+class _TimedListNode(_ListNode[P]):
```

Comment:
Could this *contain* a `_ListNode` rather than extend it? Having `_ListNode.move_after` take a `clock` is pretty ugly, and afaict the only places you call `move_after` or `insert_after` you know whether you have a `_ListNode` or a `_TimedListNode`.

Reply:
Mmm, true. Though I think extending it has a lower memory footprint:

```python
In [27]: class A(): 
    ...:     __slots__ = [] 
    ...:                                                                                                                                                                                                          

In [28]: class B(A): 
    ...:     __slots__ = [] 
    ...:                                                                                                                                                                                                          

In [29]: class C: 
    ...:     __slots__ = ["a"] 
    ...:                                                                                                                                                                                                          

In [30]: asizeof(A())                                                                                                                                                                                             
Out[30]: 16

In [31]: asizeof(B())                                                                                                                                                                                             
Out[31]: 16

In [32]: c = C()                                                                                                                                                                                                  

In [33]: c.a = A()                                                                                                                                                                                                

In [34]: asizeof(c)                                                                                                                                                                                               
Out[34]: 56
```

Reply:
True enough I guess. Does 40 bytes per entry matter that much (it seems like this is the only thing stopping `_ListNode` being a generic DLL rather than anything specific to caches)?

Reply:
I don't think its the end of the world, though I'm a bit wary we have quite a few caches that are e.g. `str -> bool`, where the overhead of metadata really starts adding up.


> Could this _contain_ a `_ListNode` rather than extend it? Having `_ListNode.move_after` take a `clock` is pretty ugly, and afaict the only places you call `move_after` or `insert_after` you know whether you have a `_ListNode` or a `_TimedListNode`.

FWIW I shuffled things around such that `_ListNode` no longer takes a clock.

Reply:
I got thoroughly nerd-sniped by this.

An object with `__slots__` needs:
 * 8 bytes for a refcount
 * 8 bytes for a pointer to the type def
 * 8 bytes for each slot
 * 16 bytes for the garbage collector (GC-able objects form a double-linked-list, so this is a pointer in each direction)

- total 40 bytes for an object with one slot.

An object with *no* slots is special-cased: it does not need to partake in the GC protocol (since it can never reference other objects), hence 16 bytes total.

A `__slots__` object which inherits from a single other `__slots__` object can do it by just adding its own slots to the end of the list of its parents slots - hence just 8 bytes for each extra slot.